File: prevoz/src/main/java/org/prevoz/android/myrides/MyRidesFragment.java
package org.prevoz.android.myrides;

import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import com.crashlytics.android.Crashlytics;
import com.timehop.stickyheadersrecyclerview.StickyRecyclerHeadersDecoration;

import org.prevoz.android.MainActivity;
import org.prevoz.android.PrevozFragment;
import org.prevoz.android.R;
import org.prevoz.android.api.ApiClient;
import org.prevoz.android.api.rest.RestRide;
import org.prevoz.android.events.Events;
import org.prevoz.android.ui.DividerItemDecoration;
import org.prevoz.android.util.PrevozActivity;
import org.prevoz.android.util.ViewUtils;

import butterknife.BindView;
import butterknife.ButterKnife;
import de.greenrobot.event.EventBus;
import retrofit.RetrofitError;
import retrofit.client.Response;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class MyRidesFragment extends PrevozFragment
{
    private static final String LOG_TAG = "Prevoz.MyRides";

    @BindView(R.id.myrides_list)
    protected RecyclerView myridesList;

    @BindView(R.id.empty_view)
    protected View emptyView;

    @BindView(R.id.myrides_throbber)
    protected ProgressBar throbber;


    private MyRidesAdapter adapter = null;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        adapter = new MyRidesAdapter(getActivity());
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        ((PrevozActivity)getActivity()).getApplicationComponent().inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View views = inflater.inflate(R.layout.fragment_myrides, container, false);
        ButterKnife.bind(this, views);
        myridesList.setHasFixedSize(true);
        myridesList.addItemDecoration(new DividerItemDecoration(getActivity(), DividerItemDecoration.VERTICAL_LIST));
        LinearLayoutManager llm = new LinearLayoutManager(getActivity(), LinearLayoutManager.VERTICAL, false);
        myridesList.setLayoutManager(llm);
        myridesList.addItemDecoration(new StickyRecyclerHeadersDecoration(adapter));
        myridesList.setAdapter(adapter);
        return views;
    }

    @Override
    public void onResume()
    {
        super.onResume();
        EventBus.getDefault().register(this);


        if (!authUtils.isAuthenticated())
        {
            authUtils.requestAuthentication(getActivity(), MainActivity.REQUEST_CODE_AUTHORIZE_MYRIDES);
        }
        else
        {
            loadRides();
        }
    }

    @Override
    public void onPause()
    {
        super.onPause();
        EventBus.getDefault().unregister(this);
    }

    private void loadRides()
    {
        ViewUtils.setupEmptyView(myridesList, emptyView, "Nimate objavljenih ali zaznamovanih prevozov.");
        setListVisibility(false);
        emptyView.setVisibility(View.INVISIBLE);
        adapter.clear();

        Observable<RestRide> myRides = ApiClient.getAdapter().getMyRides()
                .flatMap(restSearchResults -> {
                    if (restSearchResults == null || restSearchResults.results == null)
                        return Observable.empty();
                    return Observable.from(restSearchResults.results);
                });

        Observable<RestRide> bookmarks =  ApiClient.getAdapter().getBookmarkedRides()
                .flatMap(restSearchResults -> {
                    if (restSearchResults == null || restSearchResults.results == null)
                        return Observable.empty();
                    return Observable.from(restSearchResults.results);
                });

        myRides.mergeWith(bookmarks)
               .toList()
               .observeOn(AndroidSchedulers.mainThread())
                .subscribe((rides) -> {
                            adapter.addRides(rides);
                            setListVisibility(true);
                        },
                        this::showLoadFailureError);
    }

    private void showLoadFailureError(Throwable error) {
        Crashlytics.logException(error.getCause());
        if (error instanceof RetrofitError) {
            // Check for unauthorzied
            RetrofitError rerror = (RetrofitError)error;
            if (rerror.getResponse() != null) {
                Response response = rerror.getResponse();
                Crashlytics.log(Log.ERROR, LOG_TAG, response.getBody().toString());
                if (response.getStatus() == 403 || response.getStatus() == 401) {
                    authUtils.logout().subscribeOn(Schedulers.io()).toBlocking().firstOrDefault(null);
                    authUtils.requestAuthentication(getActivity(), MainActivity.REQUEST_CODE_AUTHORIZE_MYRIDES);
                    return;
                }
            }
        }

        Log.e(LOG_TAG, "Ride load failed!");
        ViewUtils.setupEmptyView(myridesList, emptyView, "Pri nalaganju vaših prevozov je prišlo do napake.");
        setListVisibility(true);
    }

    private void setListVisibility(boolean listVisible)
    {

        if (listVisible) {
            emptyView.setVisibility(adapter == null || adapter.getItemCount() == 0 ? View.VISIBLE : View.INVISIBLE);
            myridesList.setVisibility(adapter == null || adapter.getItemCount() != 0 ? View.VISIBLE : View.INVISIBLE);
        } else {
            emptyView.setVisibility(View.INVISIBLE);
            myridesList.setVisibility(View.INVISIBLE);
        }


        throbber.setVisibility(listVisible ? View.INVISIBLE : View.VISIBLE);
    }

    public void onEventMainThread(Events.MyRideStatusUpdated e)
    {
        loadRides();
    }
}

File: prevoz/src/main/kotlin/org/prevoz/android/myrides/MyRidesFragment.kt
package org.prevoz.android.myrides

import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import com.hannesdorfmann.mosby.mvp.MvpFragment
import com.timehop.stickyheadersrecyclerview.StickyRecyclerHeadersDecoration
import org.prevoz.android.PrevozApplication
import org.prevoz.android.R
import org.prevoz.android.api.rest.RestRide

class MyRidesFragment : MvpFragment<MyRidesFragment, MyRidesPresenter>() {

    @BindView(R.id.myrides_list)
    lateinit var myridesList: RecyclerView

    @BindView(R.id.empty_view)
    lateinit var emptyView: View

    @BindView(R.id.empty_text)
    lateinit var emptyText: TextView

    @BindView(R.id.empty_button)
    lateinit var loginButton: Button

    @BindView(R.id.myrides_throbber)
    lateinit var throbber: ProgressBar

    @BindView(R.id.myrides_add)
    lateinit var addButton: FloatingActionButton

    lateinit var adapter : MyRidesAdapter

    override fun createPresenter(): MyRidesPresenter {
        return MyRidesPresenter((activity.application as PrevozApplication).component())
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        adapter = MyRidesAdapter(activity)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val views = inflater?.inflate(R.layout.fragment_myrides, container, false)
        ButterKnife.bind(this, views as View)
        myridesList.setHasFixedSize(true)
        myridesList.addItemDecoration(DividerItemDecoration(activity, DividerItemDecoration.VERTICAL))
        val layoutManager = LinearLayoutManager(activity, LinearLayoutManager.VERTICAL, false)
        myridesList.layoutManager = layoutManager
        myridesList.addItemDecoration(StickyRecyclerHeadersDecoration(adapter))
        myridesList.adapter = adapter
        loginButton.text = "Prijavite se"
        return views
    }

    override fun onStart() {
        super.onStart()
        presenter.checkForAuthentication()
    }

    @OnClick(R.id.empty_button)
    fun onLoginButtonClick() {
        presenter.login()
    }

    @OnClick(R.id.myrides_add)
    fun onAddButtonClick() {
        presenter.addRide()
    }

    fun showLoginPrompt() {
        emptyText.text = "Za ogled in dodajanje zaznamkov ter prevozov morate biti prijavljeni."
        emptyView.visibility = View.VISIBLE
        myridesList.visibility = View.INVISIBLE
        throbber.visibility = View.INVISIBLE
        loginButton.visibility = View.VISIBLE
        addButton.visibility = View.INVISIBLE
    }

    fun showLoadingThrobber() {
        emptyView.visibility = View.INVISIBLE
        myridesList.visibility = View.INVISIBLE
        throbber.visibility = View.VISIBLE
        loginButton.visibility = View.GONE
        addButton.visibility = View.VISIBLE
        adapter.clear()
    }

    fun showMyRides(results: List<RestRide>) {
        adapter.setRides(results)
        myridesList.visibility = View.VISIBLE
        emptyView.visibility = View.INVISIBLE
        loginButton.visibility = View.GONE
        throbber.visibility = View.INVISIBLE
        addButton.visibility = View.VISIBLE
    }

    fun showEmptyView() {
        myridesList.visibility = View.INVISIBLE
        throbber.visibility = View.INVISIBLE
        emptyView.visibility = View.VISIBLE
        loginButton.visibility = View.GONE
        addButton.visibility = View.VISIBLE
        emptyText.text = "Nimate objavljenih ali zaznamovanih prevozov."
    }

    fun showLoadingError() {
        myridesList.visibility = View.INVISIBLE
        throbber.visibility = View.INVISIBLE
        emptyView.visibility = View.VISIBLE
        loginButton.visibility = View.GONE
        addButton.visibility = View.VISIBLE
        emptyText.text = "Pri nalaganju vaših prevozov je prišlo do napake."
    }

    fun updateRideInList(ride: RestRide) {
        adapter.updateRide(ride)
    }
}